src/counter.c

```diff
@@ -0,0 +1,98 @@
+// Commands for counting edges on GPIO input pins
+//
+// Copyright (C) 2021  Adrian Keet <arkeet@gmail.com>
+//
+// This file may be distributed under the terms of the GNU GPLv3 license.
+
+#include <stdbool.h>
+#include "basecmd.h" // oid_alloc
+#include "board/gpio.h" // struct gpio_adc
+#include "board/irq.h" // irq_disable
+#include "board/misc.h" // timer_read_time
+#include "command.h" // DECL_COMMAND
+#include "sched.h" // DECL_TASK
+
+struct counter {
+    struct timer timer;
+    struct gpio_in pin;
+    uint32_t poll_ticks;
+    uint32_t sample_ticks, last_sample_time;
+    uint32_t count, last_count_time;
+    bool value          : 1;
+    bool sample_pending : 1;
+};
+
+static struct task_wake counter_wake;
+
+static uint_fast8_t
+counter_event(struct timer *timer)
+{
+    struct counter *c = container_of(timer, struct counter, timer);
+
+    uint32_t time = c->timer.waketime;
+
+    uint8_t value = gpio_in_read(c->pin);
+    // possibly useful invariant: count % 2 == value
+    if (c->value != value) {
+        c->value = value;
+        c->count++;
+        c->last_count_time = time;
+    }
+
+    if (time - c->last_sample_time >= c->sample_ticks) {
+        c->sample_pending = true;
+        c->last_sample_time = time;
+        sched_wake_task(&counter_wake);
+    }
+
+    c->timer.waketime += c->poll_ticks;
+    return SF_RESCHEDULE;
+}
+
+void
+command_config_counter(uint32_t *args)
+{
+    struct gpio_in pin = gpio_in_setup(args[1], args[2]);
+    struct counter *c = oid_alloc(
+        args[0], command_config_counter, sizeof(*c));
+    c->timer.func = counter_event;
+    c->pin = pin;
+}
+DECL_COMMAND(command_config_counter,
+             "config_counter oid=%c pin=%u pull_up=%c");
+
+void
+command_query_counter(uint32_t *args)
+{
+    struct counter *c = oid_lookup(args[0], command_config_counter);
+    sched_del_timer(&c->timer);
+    c->timer.waketime = args[1];
+    c->poll_ticks = args[2];
+    c->sample_ticks = args[3];
+    sched_add_timer(&c->timer);
+}
+DECL_COMMAND(command_query_counter,
+             "query_counter oid=%c clock=%u poll_ticks=%u sample_ticks=%u");
+
+void
+counter_task(void)
+{
+    if (!sched_check_wake(&counter_wake))
+        return;
+
+    uint8_t oid;
+    struct counter *c;
+    foreach_oid(oid, c, command_config_counter) {
+        if (!c->sample_pending)
+            continue;
+        irq_disable();
+        uint32_t time = c->last_sample_time;
+        uint32_t count = c->count;
+        uint32_t count_time = c->last_count_time;
+        c->sample_pending = false;
+        irq_enable();
+        sendf("counter_state oid=%c time=%u count=%u count_time=%u",
+              oid, time, count, count_time);
```

Comment:
Wouldn't it be more accurate to report `c->time.waketime - c->poll_ticks` instead of `c->last_sample_time` - the `count` may have changed between the `sched_wake_task()` call and the start of this code.